sycl/include/sycl/ext/intel/experimental/esimd/simd.hpp

```diff
@@ -32,654 +41,171 @@ namespace esimd {
-    auto V2 = LHS.data() BITWISE_OP RHS.data();                                \
-    LHS.write(detail::convert<vector_type>(V2));                               \
-    return LHS;                                                                \
-  }                                                                            \
-  ESIMD_INLINE friend simd &operator OPASSIGN(simd &LHS, const Ty &RHS) {      \
-    LHS OPASSIGN simd(RHS);                                                    \
-    return LHS;                                                                \
-  }
-
-  DEF_BITWISE_OP(&, &=)
-  DEF_BITWISE_OP(|, |=)
-  DEF_BITWISE_OP(^, ^=)
-  DEF_BITWISE_OP(<<, <<=)
-  DEF_BITWISE_OP(>>, >>=)
-
-#undef DEF_BITWISE_OP
-
-  // Operator ++, --
+  /// @{
+  /// Infix and postfix operators ++, --
   simd &operator++() {
     *this += 1;
     return *this;
   }
+
   simd operator++(int) {
-    simd Ret(*this);
+    simd Ret(base_type::data());
     operator++();
     return Ret;
   }
+
   simd &operator--() {
     *this -= 1;
     return *this;
   }
+
   simd operator--(int) {
-    simd Ret(*this);
+    simd Ret(base_type::data());
     operator--();
     return Ret;
   }
+  /// @}
 
-#define DEF_UNARY_OP(UNARY_OP)                                                 \
-  simd operator UNARY_OP() {                                                   \
-    auto V = UNARY_OP(data());                                                 \
-    return simd(V);                                                            \
+#define __ESIMD_DEF_SIMD_ARITH_UNARY_OP(ARITH_UNARY_OP)                        \
+  template <class T1 = Ty> simd operator ARITH_UNARY_OP() {                    \
+    static_assert(!std::is_unsigned_v<T1>,                                     \
+                  #ARITH_UNARY_OP "doesn't apply to unsigned types");          \
+    return simd(ARITH_UNARY_OP(base_type::data()));                            \
   }
-  DEF_UNARY_OP(~)
-  DEF_UNARY_OP(+)
-  DEF_UNARY_OP(-)
-
-#undef DEF_UNARY_OP
 
-  // negation operator
-  auto operator!() { return *this == 0; }
+  __ESIMD_DEF_SIMD_ARITH_UNARY_OP(-)
+  __ESIMD_DEF_SIMD_ARITH_UNARY_OP(+)
+#undef __ESIMD_DEF_SIMD_ARITH_UNARY_OP
+};
 
-  /// \name Replicate
-  /// Replicate simd instance given a region.
-  /// @{
-  ///
+/// Covert from a simd object with element type \c From to a simd object with
+/// element type \c To.
+template <typename To, typename From, int N>
+ESIMD_INLINE simd<To, N> convert(const simd<From, N> &val) {
+  if constexpr (std::is_same_v<To, From>)
+    return val;
+  else
+    return __builtin_convertvector(val.data(), detail::vector_type_t<To, N>);
+}
 
-  /// \tparam Rep is number of times region has to be replicated.
-  /// \return replicated simd instance.
-  template <int Rep> simd<Ty, Rep * N> replicate() {
-    return replicate<Rep, N>(0);
-  }
+namespace detail {
+template <typename T, int N>
```

Comment:
Why not put this to detail folder - similar to simd_view_impl and simd_obj_impl?

Reply:
ok, makes sense